### PR TITLE
[rocksdb_env_s3] rocksdb s3 enviroment methods body implementation

### PR DESCRIPTION
**Pull request by ghanpipi**
1. We will rely on the local posix env for file opeartions: read, write,
creating, deletion etc
2. For read operations, we will first copy the files from S3 to local then
perform the read via posix_env
3. For write operations, we will first perform the write in the local tmp files,
and then copy them to s3 before closing. Due to the this s3 env will be only
used for storing backups, so this behaviour should be safe.

### CODE REVIEW

File: common/rocksdb_env_s3.cpp [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+  // we successfully copied the file, try opening it locally now
+  st = posix_env_->NewRandomAccessFile(local_full_path, result, options);
+
+  return st;
+}
+
+// S3WritableFile is a wrapper for writing a file in S3. We will perform the
+// operations in a local tmp file and upload to S3 when closing the file.
+class S3WritableFile : public WritableFile {
+ private:
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: Convention is to put the `public` section first, followed by `protected` and then finally `private`

File: common/rocksdb_env_s3.cpp [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+                 std::shared_ptr<common::S3Util> s3_util) :
+      fname_(local_fname),
+      s3_fname_(s3_fname),
+      env_(env),
+      s3_util_(s3_util) {
+    auto local_env = env_->GetBaseEnv();
+    auto s = local_env->NewWritableFile(local_fname, &local_file_, options);
+    if (!s.ok()) {
+      LOG(ERROR) << "Error happened when creating writable file in local: " << s.ToString();
+      status_ = s;
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: Any reason for setting a status in the constructor? It's odd for the caller to check status as soon as calling the constructor (and the caller might forget to do so). Would it be better to do this in a separate function?

File: common/rocksdb_env_s3.cpp [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+namespace rocksdb {
+
+// open a file for sequential reading
+Status S3Env::NewSequentialFile(const std::string& fname,
+                                std::unique_ptr<SequentialFile>* result,
+                                const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: Could the file have changed in s3 but not locally?

File: common/rocksdb_env_s3.cpp [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+  }
+
+  return Status::OK();;
+}
+
+Status S3Env::GetChildren(const std::string& path,
+                          std::vector<std::string>* result) {
+  auto s3_full_path = s3_key_prefix_ + path;
+  // first fetch all files using the given path as the key prefix in S3
+  auto resp = s3_util_->listObjects(s3_full_path);
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: If `path` does not exist, how is it handled?

File: common/rocksdb_env_s3.cpp [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+  // trim the file name
+  for (auto& v : resp.Body()) {
+    // Our path should be a prefix of the fetched value
+    if (v.find(s3_full_path) == 0) {
+      result->push_back(v.substr(s3_full_path.size()));
+    }
+  }
+
+  // then fetch all files that exist in the local posix directory
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: If you have already read from s3, what is the purpose of reading from local dir? Is it for cases when changes have been made locally but the file has not been closed?

File: common/rocksdb_env_s3.cpp [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    }
+  }
+
+  return st;
+}
+
+// The rename is not atomic. S3 does not support renaming natively.
+// Copy file to a new object in S3 and actually only IDENTITY files could be renamed.
+Status S3Env::RenameFile(const std::string& src, const std::string& target) {
+  auto resp = s3_util_->putObject(s3_key_prefix_ + target, local_directory_ + src);
```
Comment (rajathprasad) [inline common/rocksdb_env_s3.cpp]: Do we need to delete the older file?

File: common/s3util.cpp [hunk 7]
```diff
@@ -312,6 +314,22 @@ GetObjectMetadataResponse S3Util::getObjectMetadata(const string &key) {
   return GetObjectMetadataResponse(std::move(metadata), "");
 }
 
+GetObjectSizeAndModTimeResponse S3Util::getObjectSizeAndModTime(const string& key) {
+  HeadObjectRequest headObjectRequest;
+  headObjectRequest.SetBucket(bucket_);
+  headObjectRequest.SetKey(key);
+  map<string, uint64_t> metadata;
+  auto headObjectOutcome = s3Client->HeadObject(headObjectRequest);
```
Comment (newpoo) [inline common/s3util.cpp]: how is HeadObject() different from the MakeHttpRequest() call from line 297? Can we reuse getObjectMetadata()?

File: common/rocksdb_env_s3.cpp [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+Status S3Env::LockFile(const std::string& fname, FileLock** lock) {
+  // there isn's a very good way to atomically check and create
+  // a file via libs3
+  *lock = nullptr;
+  return Status::OK();
+}
+
+Status S3Env::UnlockFile(FileLock* lock) {
+  return Status::OK();
+}
```
Comment (newpoo) [inline common/rocksdb_env_s3.cpp]: what's the semantics of these functions? can you please add some comments for it. Does rocksdb backup/restore need to use these two functions?

File: common/rocksdb_env_s3.h [hunk 9]
```diff
@@ -54,7 +53,12 @@ class S3FatalException : public std::exception {
 class S3Env : public Env {
```
Comment (newpoo) [inline common/rocksdb_env_s3.h]: Can you document the design, assumption (and why), and limitation (and why) of the implementation?

File: common/rocksdb_env_s3.cpp [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+  auto local_full_path = local_directory_ + fname;
+  auto st = posix_env_->NewSequentialFile(local_full_path, result, options);
+
+  if (!st.ok()) {
+    // If file doesnt exist in local, we copy the file to the local storage from S3
+    auto resp = s3_util_->getObject(s3_key_prefix_ + fname, local_full_path, FLAGS_s3_direct_io);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when copying the file from S3 to local: "
+                 << resp.Error();
+      return Status::IOError();
```
[the remark this answers is not in the document]
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: the rocksdb will first check the status then read the value under result

File: common/rocksdb_env_s3.cpp (hunk 1 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: good catch!

File: common/rocksdb_env_s3.cpp (hunk 3 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: all the files used in the bakcup/restore are immutable.

File: common/rocksdb_env_s3.cpp (hunk 4 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: they the rocksdb backup engine wont take any operation under this path

File: common/rocksdb_env_s3.cpp (hunk 5 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: yes. we wont need to read them from local again. I already fixed it

File: common/rocksdb_env_s3.cpp (hunk 6 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: fixed in the new patch

File: common/s3util.cpp (hunk 7 above)
Comment (ghanpipi) [reply common/s3util.cpp]: some infomation like the ModTime was not exposed in the MakeHttpRequest() API

File: common/rocksdb_env_s3.h (hunk 9 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.h]: sure. Added more details.

File: common/s3util.cpp (hunk 7 above)
Comment (newpoo) [reply common/s3util.cpp]: Is "LastModified" on https://docs.aws.amazon.com/cli/latest/reference/s3api/head-object.html what we need?

File: common/rocksdb_env_s3.cpp (hunk 2 above)
Comment (ghanpipi) [reply common/rocksdb_env_s3.cpp]: I followed the convention in the rocksdb hdfs implementation

File: common/s3util.cpp (hunk 7 above)
Comment (ghanpipi) [reply common/s3util.cpp]: the format of LastModified by MakeHttpRequest is not easy to convert to the uint64_t format

**Pull request merged**